src/osd/scrubber/scrub_machine.cc

```diff
@@ -917,202 +958,32 @@ ReplicaIdle::ReplicaIdle(my_context ctx)
   dout(10) << "-- state -->> ReplicaActive/ReplicaIdle" << dendl;
 }
 
-void ReplicaIdle::reset_ignored(const FullReset&)
-{
-  dout(10) << "ReplicaIdle::react(const FullReset&): FullReset ignored"
-	   << dendl;
-}
-
-
-// ---------------- ReplicaIdle/ReplicaUnreserved ---------------------------
-
-ReplicaUnreserved::ReplicaUnreserved(my_context ctx)
-    : my_base(ctx)
-    , NamedSimply(
-	  context<ScrubMachine>().m_scrbr,
-	  "ReplicaActive/ReplicaIdle/ReplicaUnreserved")
-{
-  dout(10) << "-- state -->> ReplicaActive/ReplicaIdle/ReplicaUnreserved"
-	   << dendl;
-}
 
-sc::result ReplicaUnreserved::react(const ReplicaReserveReq& ev)
+sc::result ReplicaIdle::react(const StartReplica& ev)
 {
   DECLARE_LOCALS;  // 'scrbr' & 'pg_id' aliases
-  dout(10) << "ReplicaUnreserved::react(const ReplicaReserveReq&)" << dendl;
-
-  const auto& m = *(ev.m_op->get_req<MOSDScrubReserve>());
-  const auto async_disabled = scrbr->get_pg_cct()->_conf.get_val<bool>(
-      "osd_scrub_disable_reservation_queuing");
-  const bool async_request = !async_disabled && m.wait_for_resources;
-  dout(15) << fmt::format(
-		  "ReplicaUnreserved::react(const ReplicaReserveReq&): "
-		  "request:{} disabled?:{} -> async? {}", m.wait_for_resources,
-		  async_disabled, async_request)
-	   << dendl;
-
-  switch (context<ReplicaActive>().on_reserve_request(ev, async_request)) {
-    case ReplicaReactCode::discard:
-      return discard_event();
-    case ReplicaReactCode::goto_waiting_reservation:
-      return transit<ReplicaWaitingReservation>();
-    case ReplicaReactCode::goto_replica_reserved:
-      return transit<ReplicaReserved>();
-    default:
-      ceph_abort_msg("unexpected return value");
+  dout(10) << "ReplicaIdle::react(const StartReplica&)" << dendl;
+
+  // if we are waiting for a reservation grant from the reserver (an
+  // illegal scenario!), that reservation must be cleared.
+  if (context<ReplicaActive>().pending_reservation_nonce) {
+    scrbr->get_clog()->warn() << fmt::format(
+	"osd.{} pg[{}]: new chunk request while still waiting for "
+	"reservation",
+	scrbr->get_whoami(), scrbr->get_spgid());
+    context<ReplicaActive>().clear_remote_reservation(true);
```

Comment:
This seems pointless.  It likely won't be tested as it's impossible and it's not clear to me that clearing the reservation is actually safer.

Reply:
Clearing the reservation seems to me the safest. My reasoning:
If we are in this state, there are a few possible un-synchronized
issues that could have occurred. The most troubling is that the OSD count
(the contents of the reserver) could be off - limiting the number
of future reservations. This is also the only sync issue I can
definitely solve by clearing the reservation.

Reply:
I suppose this isn't an unreasonable choice if it actually happens, but please do not add any further complexity beyond this to support it -- focus on ensuring that it won't happen in the first place.

Reply:
This comment doesn't need to be addressed in order to merge, but the other reason to simply not check pending_reservation_nonce at all here is that it gives you the flexibility in the future to allow the primary to bypass it's reservation process mid-reservation.  There is no correctness dependency here from the replica's point of view, so it seems unnecessarily rigid to impose this requirement.

Reply:
I agree that we should not make this change now (to be backported). We should analyze & discuss that scenario. I am not sure it is the most likely reason for us to get to this reaction.

Reply:
This case does prevent the replica from leaking a reservation if the primary somehow forgot about the reservation in progress, so that's a reasonable precaution.